pkg/cdi/cache.go

```diff
@@ -65,17 +84,58 @@ func (c *Cache) Configure(options ...Option) error {
 	c.Lock()
 	defer c.Unlock()
 
+	return c.configure(options...)
+}
+
+// Configure the Cache. Start/stop CDI Spec directory watch, refresh
+// the Cache if necessary.
+func (c *Cache) configure(options ...Option) error {
+	var err error
+
 	for _, o := range options {
-		if err := o(c); err != nil {
+		if err = o(c); err != nil {
 			return errors.Wrapf(err, "failed to apply cache options")
 		}
 	}
 
-	return nil
+	if !c.selfRefresh {
+		c.watch.stop()
+		c.refresh()
+	} else {
+		err = c.watch.start(c.specDirs, &c.Mutex, c.refresh)
+	}
+
+	return err
 }
 
 // Refresh rescans the CDI Spec directories and refreshes the Cache.
+// In manual refresh mode the cache is always refreshed. In self-
+// refresh mode the cache is only refreshed if it is out of date.
 func (c *Cache) Refresh() error {
+	c.Lock()
+	defer c.Unlock()
+
+	// collect cached errors globally as a single multierror
+	cachedErrors := func() error {
+		var result error
+		for _, err := range c.errors {
+			result = multierror.Append(result, err...)
+		}
+		return result
+	}
+
+	// We need to refresh if
+	// - we're in manual refresh mode, or
+	// - in self-refresh mode a missing Spec dir appears (added to watch)
+	if !c.selfRefresh || c.watch.update() {
```

Comment:
Instead of having `selfRefresh` be a member of the cache, would it not make sense to rather instantiate a watcher that always returns `true` for `update()`?

Reply:
That would make it awkward to keep the current test/semantics in `InjectDevices`, which only `refresh()`es internally if we are in self-refresh mode and the watch got updated (missing dir created). If we removed the explicitly stored `selfRefresh` we'd need to do something like, create `watch.w` when `selfRefresh` is set, `Close()` and set it to `nil` when `selfRefresh` is cleared, and use a `watch.w != nil` test  for approximating `selfRefresh`.